chain-impl-mockchain/src/tokens.rs

```diff
@@ -0,0 +1,202 @@
+use chain_core::mempack::{ReadBuf, ReadError, Readable};
+use cryptoxide::{blake2b::Blake2b, digest::Digest};
+use thiserror::Error;
+use typed_bytes::ByteBuilder;
+
+use std::convert::TryInto;
+
+pub const POLICY_HASH_SIZE: usize = 28;
+pub const TOKEN_NAME_MAX_BYTES: usize = 32;
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct TokenIdentifier {
+    pub policy_hash: PolicyHash,
+    pub token_name: TokenName,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct PolicyHash([u8; POLICY_HASH_SIZE]);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct TokenName(Vec<u8>);
+
+/// A minting policy consists of multiple entries defining different
+/// constraints on the minting process. An empty policy means that new tokens
+/// cannot be minted during the chain run.
+///
+/// Minting policies are meant to be ignored in block0 fragments.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MintingPolicy(Vec<MintingPolicyEntry>);
```

Comment:
I think it would be nice to actually reflect this as code:
> An empty policy means that new tokens cannot be minted during the chain run.

Just as example idea:

```rust
pub enum MintingPolicy {
    Constrained(Vec<MintingPolicyEntry>)
    Locked
}
```

Reply:
How do we handle an empty `Vec` in `Constrained`?

Although one solution is to wrap an `enum`, limit its visibility to `tokens` module and make an empty `Constrained` illegal somehow.

Reply:
I somewhat agree with you, but adding an explicit enum would require the mentioned verification and it brings too much additional complexity IMO.

Reply:
I think it should be enough to wrap `Vac< MintingPolicyEntry>` into the `MintingPolicies(Vec< MintingPolicyEntry>)` (as an example) and there make an impossible to create and make an empty struct

Reply:
Which would lead to basically the same logic as we have now but with extra steps and some spaghetti code. What I like about the current structure is that:

- It is simple.
- It directly maps into its serialized representation.

Reply:
Just stop for a second and see how many layers we already invented to replace a single if-statement that did not seem nice.

Reply:
Well, I agree that the if statement may do ofc. Something like this would too https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=527415f9459e04778a03047bf4006015. 
But if it is too convoluted go ahead with the initial simple implementation.

Reply:
To me it looks a bit convoluted :)